blockchain/pool.go

```diff
@@ -1,630 +1,383 @@
+
+	// Check if the pool's maxPeerHeight has been lowered.
+	// This may happen if the tallest peer has been removed.
+	if oldMaxPeerHeight > pool.maxPeerHeight {
+		// Remove any planned requests for heights over the new maxPeerHeight.
+		for h := range pool.plannedRequests {
+			if h > pool.maxPeerHeight {
+				delete(pool.plannedRequests, h)
+			}
 		}
-
-		delete(pool.peers, peerID)
-
-		// Find a new peer with the biggest height and update maxPeerHeight if the
-		// peer's height was the biggest.
-		if peer.height == pool.maxPeerHeight {
-			pool.updateMaxPeerHeight()
+		// Adjust the nextRequestHeight to the new max plus one.
+		if pool.nextRequestHeight > pool.maxPeerHeight {
+			pool.nextRequestHeight = pool.maxPeerHeight + 1
 		}
 	}
 }
 
-// If no peers are left, maxPeerHeight is set to 0.
-func (pool *BlockPool) updateMaxPeerHeight() {
-	var max int64
+// Called every time FSM advances its height.
+func (pool *blockPool) removeShortPeers() {
 	for _, peer := range pool.peers {
-		if peer.height > max {
-			max = peer.height
+		if peer.height < pool.height {
+			pool.removePeer(peer.id, nil)
 		}
 	}
-	pool.maxPeerHeight = max
 }
 
-// Pick an available peer with at least the given minHeight.
-// If no peers are available, returns nil.
-func (pool *BlockPool) pickIncrAvailablePeer(minHeight int64) *bpPeer {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
+func (pool *blockPool) removeBadPeers() {
+	pool.removeShortPeers()
 	for _, peer := range pool.peers {
-		if peer.didTimeout {
-			pool.removePeer(peer.id)
-			continue
-		}
-		if peer.numPending >= maxPendingRequestsPerPeer {
-			continue
+		if err := peer.isGood(); err != nil {
+			pool.removePeer(peer.id, err)
+			peer.errFunc(err, peer.id)
 		}
-		if peer.height < minHeight {
-			continue
-		}
-		peer.incrPending()
-		return peer
 	}
-	return nil
 }
 
-func (pool *BlockPool) makeNextRequester() {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
-	nextHeight := pool.height + pool.requestersLen()
-	if nextHeight > pool.maxPeerHeight {
-		return
-	}
-
-	request := newBPRequester(pool, nextHeight)
-
-	pool.requesters[nextHeight] = request
-	atomic.AddInt32(&pool.numPending, 1)
-
-	err := request.Start()
-	if err != nil {
-		request.Logger.Error("Error starting request", "err", err)
+// Make a batch of requests sorted by height. The parameter 'maxNumRequests' includes the number of block requests
+// already made.
+func (pool *blockPool) makeRequestBatch(maxNumRequests int32) []int {
+	pool.removeBadPeers()
+	// At this point pool.requests may include heights for requests to be redone due to removal of peers:
+	// - peers timed out or were removed by switch
+	// - FSM timed out on waiting to advance the block execution due to missing blocks at h or h+1
+	// Check if more requests should be tried by subtracting the number of requests already made from the maximum allowed
+	numNeeded := int(maxNumRequests) - len(pool.blocks)
+	for len(pool.plannedRequests) < numNeeded {
```

Comment:
I find hard understanding this logic. I guess part of the problem might be the fact that we use `len(pool.blocks)` and `len(pool.plannedRequests)` as something semantic and it's not obvious what it corresponds to.